docs/proposal-0007-discrete_metric_execution.md

```diff
@@ -0,0 +1,107 @@
+
+
+## Decision Drivers
+
+* Ability to scale without concern for number of metrics
+* Reduce manual support needed for automated analysis errors (reduce/eliminate errors)
+* Enable (or not inhibit) known upcoming features
+* Ability to (re-)run individual metrics
+
+### Non-Drivers
+
+* We are not specifically setting **cost reduction** as a goal of these changes. We will, however, be monitoring for any cost changes that result. Additionally, something we do not want is for the proposed solution to include significantly increased costs as an *expectation*.
+* Similarly, we are not setting **total execution time** as a goal of these changes. The scalability decision driver is similar to this, but what we care about is that we can scale without failing, not without increasing total execution time.
+
+
+## Description of Solution
+
+The chosen solution involves a few primary components that will each contribute to some or all of the decision drivers.
+
+### Discrete Metric Execution
+
+For each metric (parallel execution):
+* query metric from telemetry (we can aggregate all metrics for a given data source)
+* write to metric table with results (delete existing **column data** first (currently we delete and re-write the entire table with one query))
+* compute statistics for the metric
+* write to statistics table with results (behavior unchanged)
+
+#### Questions
+* query for every metric or for every metric *data source*?
+  * data source, but have CLI option to execute single metric at a time
+
+#### Pros / Cons
+
+* **+** Adding more metrics does not increase the complexity of any single query
+* **+** Ability to execute individual metrics as CLI parameter (e.g., for rerun)
+* **+** Individual metric error does not need to cause entire job to fail
+* **-** More total queries (by factor of `m` where `m` is the # of metrics (or # of metric data sources, possibly))
+
+#### Query Analysis
+
+I picked a recent experiment, and ran a single metrics query with 4 data sources, and then 4 discrete metrics queries, one for each data source (full query and data [here](https://docs.google.com/document/d/1wcHTsnG75oeaABWswcnu_nv4RfKf5647LlaClJFi84c/edit#heading=h.73nlojxheztt)). Here are the takeaways:
+
+For the single query vs sum of four queries:
+* Total bytes processed (per BigQuery console before execution) was the same
+* Bytes shuffled was the same
+* Elapsed time was ~33% greater for the 4 queries
+* Slot time was ~50% less for the 4 queries
+
+
+### Analysis Period Dependencies
+
+The idea of this component is to define dependencies between analysis periods so that they can be computed in a particular order. This would allow us to:
+* Compute multi-day analysis periods based off of single day analysis period results instead of re-computing from scratch
+* Compute pre-enrollment periods before post-enrollment periods (for covariate adjustment)
+
+#### Questions
+* should this be configurable? if so, configured by experiment, app, etc.?
```

Comment:
I'm not sure. I think the set of analysis periods and the dependencies between them is pretty static, it's not clear to me how often that would need to be changed

Reply:
Sounds good. I'll see if there's an easy opportunity to make it configurable, but won't worry about it too much.